### PR TITLE
Implicitly add outputs to modules when they are being used across deployment groups

### PR DESCRIPTION
**Pull request by tpdownes**
This PR builds upon the same branch being merged in #877. The last 3 commits are the new additions. These commits...

- modify `getModuleByID` to explicitly error rather than rely upon the user to know how to detect an error in the output
- when intergroup references are made explicitly via module `settings` or `use` keywords, the used module's outputs are implicitly modified to include the targeted output (does not trigger when intragroup references are made)
- a final commit that enables blueprints to be written without generating warnings during `ghpc create`

This will ultimately faciliate the creation of blueprints with intergroup references via Toolkit variables and the `use` keyword.

```
---
blueprint_name: igc

vars:
  project_id: ####
  deployment_name: test
  region: us-east4
  zone: us-east4-c

deployment_groups:
- group: zero
  modules:
  - id: network0
    source: modules/network/vpc
  - id: homefs
    source: modules/file-system/filestore
    use: [network0]
    settings:
      local_mount: /home
  - id: projectsfs
    source: modules/file-system/filestore
    use: [network0]
    settings:
      local_mount: /projects
- group: one
  modules:
  - id: image
    source: modules/packer/custom-image
    kind: packer
    use:
    - zero.network0
- group: two
  modules:
  - id: network1
    source: modules/network/pre-existing-vpc
    use:
    - zero.network0
  - id: low_cost_partition
    source: community/modules/compute/SchedMD-slurm-on-gcp-partition
    use:
    - network1
    - zero.homefs
    - zero.projectsfs
    settings:
      partition_name: low_cost
      max_node_count: 10
      enable_placement: false
      exclusive: false
      machine_type: n2-standard-4
```

### Submission Checklist

* [x] Have you installed and run this change against pre-commit? (`pre-commit
  install`)
* [x] Are all tests passing? (`make tests`)
* [x] Have you written unit tests to cover this change?
* [x] Is unit test coverage still above 80%?
* [x] Have you updated all applicable documentation?
* [x] Have you followed the guidelines in our Contributing document?

### CODE REVIEW

File: pkg/config/expand.go [hunk 1]
```diff
@@ -710,11 +711,15 @@ func expandSimpleVariable(
...
-			// TODO: expandedVariable = fmt.Sprintf("((var.%s_%s))", ref.Name, ref.ID)
-			return "", fmt.Errorf("%s: %s is an intergroup reference",
-				errorMessages["varInAnotherGroup"], context.varString)
+			switch fromMod.Kind {
+			case "terraform":
+				expandedVariable = fmt.Sprintf("((var.%s_%s))", ref.Name, ref.ID)
+			case "packer":
+				expandedVariable = "((TEST))"
+			}
+		} else {
```
Comment (nick-stroud) [inline pkg/config/expand.go:720]: I would put the simple intragroup case first. It keeps the if -> else closer together and more readable.

File: pkg/config/expand.go [hunk 2]
```diff
@@ -710,11 +711,15 @@ func expandSimpleVariable(
...
 				toMod.Outputs = append(toMod.Outputs, ref.Name)
 			}
-			// TODO: expandedVariable = fmt.Sprintf("((var.%s_%s))", ref.Name, ref.ID)
-			return "", fmt.Errorf("%s: %s is an intergroup reference",
-				errorMessages["varInAnotherGroup"], context.varString)
+			switch fromMod.Kind {
+			case "terraform":
+				expandedVariable = fmt.Sprintf("((var.%s_%s))", ref.Name, ref.ID)
+			case "packer":
+				expandedVariable = "((TEST))"
```
Comment (nick-stroud) [inline pkg/config/expand.go:718]: I would have expected error on this case. I am betting we error somewhere else?

File: pkg/config/config.go [hunk 3]
```diff
@@ -96,14 +97,12 @@ type DeploymentGroup struct {
...
-func (g DeploymentGroup) getModuleByID(modID string) Module {
-	for i := range g.Modules {
-		mod := g.Modules[i]
-		if g.Modules[i].ID == modID {
-			return mod
-		}
+func (g DeploymentGroup) getModuleByID(modID string) (Module, error) {
+	idx := slices.IndexFunc(g.Modules, func(m Module) bool { return m.ID == modID })
+	if idx == -1 {
+		return Module{}, fmt.Errorf("%s: %s", errorMessages["invalidMod"], modID)
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/config.go:104]: `nil` would only be a valid value if we were returning a pointer `*Module` or other pointer-like thing (e.g. a slice).

File: pkg/config/expand.go [hunk 4]
```diff
@@ -164,6 +165,7 @@ func getModuleInputMap(inputs []modulereader.VarInfo) map[string]string {
 func useModule(
 	mod *Module,
 	useMod Module,
+	useModGroupID string,
 	modInputs map[string]string,
 	useOutputs []modulereader.VarInfo,
 	settingsInBlueprint []string,
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/expand.go:171]: There is a lot of overlap of terms here, some of it avoidable, but I understand the confusion. TL;DR: yes, that is right.

There is the `Module` struct in the blueprint with `Settings`. Then there is the actual Terraform module represented by the `modulereader.ModuleInfo` struct with `Inputs` and `Outputs`. It is probably right not to have this second set of info in an object that is part of the blueprint.

In short, we apply `Settings` to `Inputs`. So we do need the 2 separate sets of info. I think the code might have been more easily read if the inputs and objects arguments were both of type `VarInfo` in the first place.

[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/expand.go:171]: No we cannot because `useModule` is being called iteratively over every element of the `use` list in `applyUseModules` below. This will (for better/worse) iteratively modify `mod.Settings` in-place. One of the mechanisms by which it modifies `mod.Settings` is to build up a list  when the corresponding `Input` is of type `list`.

A more accurate name for this variable may be `settingsToIgnore` with the clarification about expectation that it would be used to ignore explicit settings in the blueprint.

File: pkg/config/expand.go [hunk 5]
```diff
@@ -176,14 +178,16 @@ func useModule(
...
-		// ensuring that the first used module takes precedence and also
-		// when flattening multiple values for settings that are lists
+		// if the input exists in the module doing the using, then we modify it
+		// according to these rules:
+		// if the setting has a value, then it was set by "use" because we have
+		// skipped anything explicitly in the blueprint. If the setting is a
+		// list of any type, then we want to concatenate these values in order
+		// they were used; otherwise, the setting should remain set to the
+		// existing used value
 		_, isAlreadySet := mod.Settings[settingName]
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/expand.go:188]: Strictly speaking this would be true for how the function ought to be used, but it might not be the case if one supplies an empty `settingsInBlueprint`. I'll consider a change for clarification.

File: pkg/config/expand.go [hunk 6]
```diff
@@ -218,12 +222,27 @@ func (dc *DeploymentConfig) applyUseModules() error {
 			fromModInputs := getModuleInputMap(fromModInfo.Inputs)
 			settingsInBlueprint := maps.Keys(fromMod.Settings)
 			for _, toModID := range fromMod.Use {
-				toMod, err := group.getModuleByID(toModID)
+				modRef, err := identifyModuleByReference(toModID, *group)
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/expand.go:225]: I will improve the commenting here.

File: pkg/config/expand.go [hunk 7]
```diff
@@ -680,6 +699,17 @@ func expandSimpleVariable(
...
+			// call to function that adds outputs in ToGroupID
+			toGrpIdx := modToGrp[ref.ToGroupID]
+			toModIdx := slices.IndexFunc(context.blueprint.DeploymentGroups[toGrpIdx].Modules, func(m Module) bool { return m.ID == ref.ID })
+			if toModIdx == -1 {
+				return "", fmt.Errorf("%s: %s", errorMessages["invalidMod"], ref.ID)
+			}
+			toMod := &context.blueprint.DeploymentGroups[toGrpIdx].Modules[toModIdx]
+
+			if !slices.Contains(toMod.Outputs, ref.Name) {
+				toMod.Outputs = append(toMod.Outputs, ref.Name)
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/expand.go:712]: I consider that future work (maybe short-term future) and requires changes to the `modulereader.ModuleInfo` object to implement. I think that will be a small PR once this is equivalent.

File: pkg/config/expand.go [hunk 8]
```diff
@@ -710,11 +711,15 @@ func expandSimpleVariable(
 			if !slices.Contains(toMod.Outputs, ref.Name) {
 				toMod.Outputs = append(toMod.Outputs, ref.Name)
 			}
-			// TODO: expandedVariable = fmt.Sprintf("((var.%s_%s))", ref.Name, ref.ID)
-			return "", fmt.Errorf("%s: %s is an intergroup reference",
-				errorMessages["varInAnotherGroup"], context.varString)
+			switch fromMod.Kind {
+			case "terraform":
+				expandedVariable = fmt.Sprintf("((var.%s_%s))", ref.Name, ref.ID)
```
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/expand.go:716]: Can do.

File: pkg/config/expand.go (hunk 2 above)
Comment (tpdownes) [reply pkg/config/expand.go:718]: This code is a proxy for the future functionality of adding corresponding inputs. I believe this block is eventually going to modify a struct that tracks "intergroup data" to collect from previous groups to guide the user (or automated solution) on how to apply values from previous groups. This will be true for both Packer and Terraform. Because of that, it probably will not error for Packer and we will have to return some sort of valid string value that will not be written to disk in `defaults.auto.pkrvars.hcl`. I am considering the empty Toolkit literal `"(())"`.

File: pkg/config/expand.go (hunk 1 above)
Comment (tpdownes) [reply pkg/config/expand.go:720]: Will do.

File: pkg/config/expand.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/expand.go:188]: I have changed it to `setByUse`

File: pkg/config/expand.go (hunk 1 above)
Comment (tpdownes) [reply pkg/config/expand.go:720]: go-critic complained about the revised solution and advised using a `switch`. please see update with comments.

File: pkg/config/expand.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (nick-stroud) [reply pkg/config/expand.go]: SG.

File: pkg/config/expand.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nick-stroud) [reply pkg/config/expand.go:171]: nit: I think here and in other comments you have some constructive ideas on how to change the signature of this function. I know it is not directly needed for your work, but I would strongly ask/recommend that in a follow on pr you refactor the signature of this function with your suggestions.

File: pkg/config/expand.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (nick-stroud) [reply pkg/config/expand.go:235]: Nit: PS: I think commenting in its final form is is too verbose. Will share reference separately. I was more thinking of some white space and a single line comment in this specific location. It is probably a sign that the code has become two dense and needs some refactoring or breaking out into separate functions.

File: pkg/config/expand.go (hunk 1 above)
Comment (nick-stroud) [reply pkg/config/expand.go]: looks better.

File: pkg/config/expand.go (hunk 2 above)
Comment (nick-stroud) [reply pkg/config/expand.go]: Ahh. This is a packer module referencing a terraform module. I was thinking the other way around. Correct? If so then it makes sense to me why we wouldn't error.

I am still not sure why you would leave it empty. Would this not be some variable name that we are feeding into Packer. In my current understanding I would expect it to look similar to the Terraform reference. Maybe I am missing something critical about how Packer does variables.

File: pkg/config/expand.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (nick-stroud) [reply pkg/config/expand.go]: Nit: Can you add a task for this or add a comment to an existing task. As completion of this is not required for mvp and this was explicitly discussed by group, I want to make sure we don't lose track of it.

[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/expand.go]: I can do that. Part of the reason for not sneaking it in is that I think we need to consider also sensitive inputs and other mechanisms by which to reduce the output of `terraform` commands.

File: pkg/config/expand.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/expand.go:235]: I agree that the verbosity is a sign that this area of the code needs a rethink. I aim to document lessons here with (at least) toy examples of how it could be improved conceptually and technically. The big problem here is that the values stay "raw" (e.g. strings) even after they have been validated.

File: pkg/config/expand.go (hunk 2 above)
Comment (tpdownes) [reply pkg/config/expand.go]: Yes, this covers the use case of Packer using Terraform (e.g. a subnetwork created by VPC module with known firewall properties). The critical differences between Packer "templates" and Terraform modules, which we gloss over by calling each Toolkit modules:

1. Packer has no concept of state
1. Packer has no native concept of outputs
  * thus we cannot "use" Packer modules
  * there is a task to define a Toolkit notion of Packer outputs (specifically to get image_name)
1. Packer has no mechanism by which to refer to a remote template (i.e. a Packer template has no child templates)

(3) means we are stuck with the input variables present in the Packer template itself and cannot construct new ones like we do for the Terraform root modules. There is a later section of code (in modulewriter) that writes `default.auto.prkvals.hcl` which is the equivalent of `terraform.tfvars`. This code will attempt to evaluate anything with pattern `((var.variable_name))` in the "context" of our deployment variables (`vars.*`) so that it imitates what we do for `terraform.tfvars`.

We should evaluate whether the better place to do that is in this section of code in the first place. This is a tough "[traversal](https://pkg.go.dev/github.com/hashicorp/hcl2/hcl#Traversal)" through the code. I will create a task do that.

Either way, for intergroup references, we need to have something here that indicates the variable should not be written to `defaults.auto.pkrvals.hcl` at all and needs to be collected (by user or by ghpc) from prior Terraform state before running `packer build`.

Hopefully this essay clarifies the situation.

File: pkg/config/expand.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (cboneti) [reply pkg/config/expand.go:171]: In this PR, I would strongly suggest that you look at cyclomatic complexity as a proxy for code quality and then keep it <= 7. That should force you to rethink this code. For instance, I feel this function is now way too complex.

[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/expand.go:171]: The cyclomatic complexity of this function is 7 before and after the PR.  During PR review, we were able to make a change that dedented by the block by "exiting early" (a 1-for-1 change in cyclomatic complexity but hopefully more conceptually clear and readable).

That said, while developing these changes, I gave consideration to altering the block that now reads... `if isInputList { ... else if !setByUse ... }`. In broad strokes, I considered:

- a refactor that would put `if setByUse` as the outer block
- improving commenting and variable naming for clarity

I concluded that improving the variable names and commenting was the best alternative since it would be least likely to introduce unintentional subtle behavioral changes (i.e. a mistake). The ultimate complexity here stems from the fact that our `Settings` are just `interface{}` and we need to initialize settings whose input types are lists to ` []interface{}{}` and then append to them rather than overwrite them.

I can see value in a `func` that initializes a setting of type list. I'm not sure it will reduce the cyclomatic complexity but it is something that ought to have one correct implementation.

[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/expand.go:171]: I think the long comment might be a better fit as a comment to the func itself rather than to the block.

LMK.

File: pkg/config/expand.go (hunk 2 above)
Comment (nick-stroud) [reply pkg/config/expand.go]: I think I understand conceptually but am still struggling to see why this leads us to have `(())`. 

IIUC this was part of the commit that was popped off of the top, correct? If so feel free to close this comment and we can discuss further later.

Comment (tpdownes) [reply pkg/config/expand.go]: That is correct. This change is not part of the PR and will be part of the subsequent work to glue outputs and inputs together across groups.

File: pkg/config/expand.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nick-stroud) [reply pkg/config/expand.go:171]: If you are proposing moving all the comments into a single block at the top, I don't know if that is better. My advice would be to merge as is and then spend 4 hours to make this section incrementally better in one or more separate prs. My suggestion would be:
- Break out a few functions to make the code self documenting
- Change the function signature with the ideas you have proposed
- Add some minimal function descriptions that briefly describe the purpose of the functions

**Review by nick-stroud — APPROVED** [review]

**Comment by cboneti** [comment]
I won't have time to review, and if you and Nick agree it is all good, I don't need to. :) 
Assigned to @tpdownes

File: pkg/config/expand.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tpdownes) [reply pkg/config/expand.go:171]: I agree with these changes in broad strokes and will submit some/all of them as a separate PR.
